### PR TITLE
Added support for non-docker container runtime, e.g. containerd or cri-o

### PR DESCRIPTION
**Pull request by zhouzijiang**
#### Type of fix: Bug Fix

Please describe:
- changes made in the Pull request

when kubelet creates a pod, it will call RunPodSandbox function:

1. when the container runtime is docker in kubelet, it has the following steps:
    1.1 create the sandbox container(it will be in the new network namespace)
    1.2 start the sandbox container
    1.3 kubelet forks and execs binary contivk8s and pass CNI_NETNS like '/proc/37544/ns/net'
    1.4 contivk8s pass '/proc/37544/ns/net' to netplugin
    1.5 netplugin get the new network namespace from '/proc/37544/ns/net'
    1.6 netplugin execs command like 'nsenter -t 37544 -n ip address add 192.168.1.1/24 dev eth0'

2. when the container runtime is containerd or other in kubelet, it has the following steps:
    2.1 containerd saves the current network namespace
    2.2 containerd calls unshare to move it to a new network namespace
    2.3 containerd creates file like '/var/run/netns/cni-eb58672e-a27e-ac5d-73dd-7ebb56282818' and bind mounts it to the new network namespace
    2.4 containerd restores itself to the original network namespace
    2.5 now there are no processes in the new network namespace, and we can only access the new network namespace by file '/var/run/netns/cni-eb58672e-a27e-ac5d-73dd-7ebb56282818'
    2.6 containerd forks and execs binary contivk8s and pass CNI_NETNS like '/var/run/netns/cni-eb58672e-a27e-ac5d-73dd-7ebb56282818'
    2.7 contivk8s pass '/var/run/netns/cni-eb58672e-a27e-ac5d-73dd-7ebb56282818' to netplugin
    2.8 netplugin can not parse '/var/run/netns/cni-eb58672e-a27e-ac5d-73dd-7ebb56282818' and return error

So, our solution is changing step 2.7:
1. contivk8s gets CNI_NETNS like '/var/run/netns/cni-eb58672e-a27e-ac5d-73dd-7ebb56282818'
2. contivk8s gets the new network namespace from '/var/run/netns/cni-eb58672e-a27e-ac5d-73dd-7ebb56282818'
3. contivk8s makes itself to the new network namespace
4. contivk8s creates the new argument by code 'fmt.Sprintf("/proc/%d/ns/net", os.Getpid())'
5. contivk8s passes '/proc/12345/ns/net' to netplugin

- type of testing done (both manual and automated)

## TODO
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by zhouzijiang** [comment]
@tiewei @vhosakot please review it, thanks again

File: mgmtfn/k8splugin/contivk8s/k8s_cni.go [hunk 1]
```diff
@@ -63,6 +64,21 @@ func getPodInfo(ppInfo *cniapi.CNIPodAttr) error {
 
 	// nwNameSpace and ifname are passed as separate env vars
 	ppInfo.NwNameSpace = os.Getenv("CNI_NETNS")
+	if ppInfo.NwNameSpace != "" {
+		ok := strings.HasPrefix(ppInfo.NwNameSpace, "/proc/")
```
Comment (tiewei) [inline mgmtfn/k8splugin/contivk8s/k8s_cni.go]:
```suggestion
		if !strings.HasPrefix(ppInfo.NwNameSpace, "/proc/")
```

File: mgmtfn/k8splugin/contivk8s/k8s_cni.go [hunk 2]
```diff
@@ -63,6 +64,21 @@ func getPodInfo(ppInfo *cniapi.CNIPodAttr) error {
...
+		if !ok {
+			nsHandle, err := netns.GetFromPath(ppInfo.NwNameSpace)
+			if err != nil {
+				return fmt.Errorf("error getting ns %s: %s", ppInfo.NwNameSpace, err)
+			}
+
+			if err = netns.Set(nsHandle); err != nil {
+				return fmt.Errorf("error switching to ns %s: %s", ppInfo.NwNameSpace, err)
+			}
```
Comment (tiewei) [inline mgmtfn/k8splugin/contivk8s/k8s_cni.go]:
```suggestion
			if nsHandle, err := netns.GetFromPath(ppInfo.NwNameSpace); err != nil {
			   return fmt.Errorf("error getting ns %s: %s", ppInfo.NwNameSpace, err)
			} else if err = netns.Set(nsHandle); err != nil { 
			    return fmt.Errorf("error switching to ns %s: %s", ppInfo.NwNameSpace, err)
			}
			
```

File: mgmtfn/k8splugin/contivk8s/k8s_cni.go [hunk 3]
```diff
@@ -152,10 +168,14 @@ func getPrefixedLogger() *logger.Entry {
 	var nsID string
 
 	netNS := os.Getenv("CNI_NETNS")
-	ok := strings.HasPrefix(netNS, "/proc/")
-	if ok {
-		elements := strings.Split(netNS, "/")
-		nsID = elements[2]
+	if netNS != "" {
+		ok := strings.HasPrefix(netNS, "/proc/")
+		if ok {
```
Comment (tiewei) [inline mgmtfn/k8splugin/contivk8s/k8s_cni.go]:
```suggestion
		if ok := strings.HasPrefix(netNS, "/proc/"); ok {
```

**Comment by zhouzijiang** [comment]
> Thanks for submitting reviews, the logic here sound, do you mind adding some test cases ?

ok, I will add some test cases. In fact, we have run contiv with this feature in our k8s cluster for about a year

**Comment by zhouzijiang** [comment]
@tiewei  I have added some test cases

**Comment by cusri-admin** [comment]
very good

**Pull request merged**